objectherkenning_openbare_ruimte/convert_annotations_pipeline/source/yolo_to_azure_coco_converter.py

```diff
@@ -0,0 +1,196 @@
+import json
+import os
+from datetime import datetime
+from typing import Any, Dict
+
+import cv2
+from cvtoolkit.helpers.file_helpers import find_image_paths  # noqa: E402
+
+
+class YoloToAzureCocoConverter:
+    """
+    Converts a YOLO annotation dataset to Azure COCO format.
+    """
+
+    def __init__(
+        self,
+        input_folder: str,
+        output_folder: str,
+        datastore_name: str,
+        categories_file: str,
+    ):
+        """
+        Parameters
+        ----------
+        input_folder: str
+            Path to the folder containing the YOLO annotations.
+        output_folder: str
+            Path to the folder where the Azure COCO annotations will be stored.
+        datastore_name: str
+            Name of the datastore to be used in the COCO file URLs.
+        categories_file: str
+            Path to the JSON file containing the categories.
+        """
+        self.input_folder = input_folder
+        self.output_folder = output_folder
+        self.datastore_name = datastore_name
+        self._load_categories(categories_file)
+        self.coco_json = {
+            "images": [],
+            "annotations": [],
+            "categories": self.categories,
+        }
+        self.annotation_id = 1
+
+    def _load_categories(self, categories_file: str):
+        """
+        Loads categories from a JSON file.
+
+        Parameters
+        ----------
+        categories_file: str
+            Path to the JSON file containing the categories.
+        """
+        with open(categories_file, "r") as f:
+            self.categories = json.load(f)
+
+    @staticmethod
+    def _parse_image_path(img_path: str) -> tuple:
```

Comment:
This looks a bit hacky, is there not a better way?
Was there not a function for this already somewhere?

Reply:
It is indeed a bit hacky. I already use `find_image_paths`, but it outputs a path with this format:

- `/mnt/azureml/[...]/wd/` + any folders before the last one containing the images + `/INPUT_input_old_folder/` + `images_folder/image_name.jpg` or `image_name.jpg`.

With `_parse_image_path`, we are robust to any folder structure (for example, the code works both in the case there is a structure like `/images/test/test2/images_folder/` and in the case it's only `/images_folder/`.

An alternative could be to assert that there are no other folders before `image_folder`, so the function would just be: `folder_name = img_path.split("/")[-2]`.